// testdouble/lib/store/index.js
// Generated by CoffeeScript 1.11.1
(function() {
  var EventEmitter, _, globalStore, initialEntryFor, storeEmitter;

  _ = require('../util/lodash-wrap');

  EventEmitter = require('events').EventEmitter;

  storeEmitter = new EventEmitter();

  globalStore = [];

  module.exports = {
    onReset: function(func) {
      return storeEmitter.on('reset', func);
    },
    reset: function() {
      globalStore = [];
      return storeEmitter.emit('reset');
    },
    "for": function(testDouble, createIfNew) {
      var entry;
      if (createIfNew == null) {
        createIfNew = true;
      }
      if (entry = _.find(globalStore, {
        testDouble: testDouble
      })) {
        return entry;
      }
      if (!createIfNew) {
        return;
      }
      return _.tap(initialEntryFor(testDouble), function(newEntry) {
        return globalStore.push(newEntry);
      });
    }
  };

  initialEntryFor = function(testDouble) {
    return {
      testDouble: testDouble,
      stubbings: [],
      calls: [],
      verifications: []
    };
  };

}).call(this);
